web/src/user_group_pill.ts

```diff
@@ -1,12 +1,16 @@
 import type {InputPillContainer, InputPillItem} from "./input_pill";
+import type { CombinedPillContainer } from "./pill_typeahead";
 import type {UserGroup} from "./user_groups";
 import * as user_groups from "./user_groups";
 
-type UserGroupPill = {
+export type UserGroupPill = {
+    type: "user_group",
     id: number;
     group_name: string;
 };
 
+export type UserGroupPillData = UserGroup & {type: "user_group"};
```

Comment:
Do you need the `type` field both on the PillData object and on the `Pill` itself? I don't really know how this works.

Reply:
The type for the pill is for functions that deal with `CombinedPillContainer`, like:

```js
export function get_group_ids(pill_widget: CombinedPillContainer): number[] {
    const items = pill_widget.items();
    return items.flatMap((item) => item.type === "user_group" ? item.id : []);
}
```

The type for the pill data is for functions that deal with `TypeaheadItem`, like:

```js
        highlighter_html(item: TypeaheadItem, query: string): string {
            if (include_streams(query) && item.type === "stream") {
                return typeahead_helper.render_stream(item);
            }

            if (include_user_groups && item.type === "user_group") {
                return typeahead_helper.render_user_group(item);
            }
```

They're just two different parts of the pill typeahead that use union types of these different kinds of pills, and I'm pretty sure they both need the `type` field.

Reply:
OK, certainly seems fine if it's necessary.